File: array/leetcode_24.java
package array;

public class leetcode_24 {
    public static void main(String[] args) {
        leetcode_24 lc = new leetcode_24();
        ListNode list = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
        System.out.println(lc.swapPairs(list));
    }

    private static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode swapPairs(ListNode head) {
        ListNode list = head;

        while (head != null && head.next != null) {
            swap(head, head.next);
            head = head.next.next;
        }

        return list;
    }

    private static void swap(ListNode node1, ListNode node2) {
        int tmp = node1.val;
        node1.val = node2.val;
        node2.val = tmp;
    }
}

File: linked_list/LeetCode_24.kt
package linked_list

import util.ListNode

class LeetCode_24 {

    fun swapPairs(head: ListNode?): ListNode? {
        if (head == null) {
            return null
        }

        val headIndex = ListNode(-1)

        var clonedHead = head

        headIndex.next = clonedHead

        var idx = headIndex

        while (clonedHead?.next != null) {
            val tmp = clonedHead
            clonedHead = clonedHead.next
            tmp.next = clonedHead!!.next
            clonedHead.next = tmp

            idx.next = clonedHead

            clonedHead = tmp.next
            idx = idx.next!!.next!!
        }

        return headIndex.next
    }
}

fun main() {
    // Input
    val head = ListNode.create(1, 2, 3, 4)
    // Output
    LeetCode_24().swapPairs(head).run {
        println(this)
        println(this?.equals(ListNode.create(2, 1, 4, 3)))
    }
}
